python/distributed-ucxx/distributed_ucxx/ucxx.py

```diff
@@ -90,13 +91,102 @@ def synchronize_stream(stream=0):
     stream.synchronize()
 
 
+def make_register():
+    """Register a Dask resource with the UCXX context.
+
+    Register a Dask resource with the UCXX context and keep track of it with the
+    use of a unique ID for the resource. The resource ID is later used to
+    deregister the resource from the UCXX context calling
+    `_deregister_dask_resource(resource_id)`, which stops the notifier thread
+    and progress tasks when no more UCXX resources are alive.
+
+    Returns
+    -------
+    resource_id: int
+        The ID of the registered resource that should be used with
+        `_deregister_dask_resource` during stop/destruction of the resource.
+    """
+
+    count = itertools.count()
+
+    def register() -> int:
+        ctx = ucxx.core._get_ctx()
+        with ctx._dask_resources_lock:
+            resource_id = next(count)
+            ctx._dask_resources.add(resource_id)
+            ctx.start_notifier_thread()
+            ctx.continuous_ucx_progress()
+            return resource_id
+
+    return register
+
+
+_register_dask_resource = make_register()
+
+
+def _deregister_dask_resource(resource_id):
+    """Deregister a Dask resource with the UCXX context.
+
+    Deregister a Dask resource from the UCXX context with given ID, and if no
+    resources remain after deregistration, stop the notifier thread and progress
+    tasks.
+
+    Parameters
+    ----------
+    resource_id: int
+        The unique ID of the resource returned by `_register_dask_resource` upon
+        registration.
+    """
+    if ucxx.core._ctx is None:
+        # Prevent creation of context if it was already destroyed, all
+        # registered references are already gone.
+        return
+
+    ctx = ucxx.core._get_ctx()
+
+    # Check if the attribute exists first, in tests the UCXX context may have
+    # been reset before some resources are deregistered.
+    if hasattr(ctx, "_dask_resources_lock"):
+        with ctx._dask_resources_lock:
+            try:
+                ctx._dask_resources.remove(resource_id)
+            except KeyError:
+                pass
```

Comment:
That's all correct. I've also relied on the unique values of resource references/hashes for debugging, so that was more useful for debugging like that. As you say, I also don't think it's worth it but may be done if we find a strong reason to follow that path.